executor/importer/chunk_process.go

```diff
@@ -0,0 +1,300 @@
+	"github.com/pingcap/tidb/br/pkg/lightning/checkpoints"
+	"github.com/pingcap/tidb/br/pkg/lightning/common"
+	"github.com/pingcap/tidb/br/pkg/lightning/log"
+	"github.com/pingcap/tidb/br/pkg/lightning/mydump"
+	verify "github.com/pingcap/tidb/br/pkg/lightning/verification"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/tikv/client-go/v2/tikv"
+	"go.uber.org/zap"
+)
+
+// constants, make it a variable for test
+var (
+	maxKVQueueSize         = 32             // Cache at most this number of rows before blocking the encode loop
+	MinDeliverBytes uint64 = 96 * units.KiB // 96 KB (data + index). batch at least this amount of bytes to reduce number of messages
+	// see default for tikv-importer.max-kv-pairs
+	MinDeliverRowCnt = 4096
+)
+
+type deliveredRow struct {
+	kvs    *kv.Pairs // if kvs is nil, this indicated we've got the last message.
+	offset int64
+	rowID  int64
+}
+
+type deliverResult struct {
+	totalDur time.Duration
+	err      error
+}
+
+type deliverKVBatch struct {
+	dataKVs  kv.Pairs
+	indexKVs kv.Pairs
+
+	dataChecksum  *verify.KVChecksum
+	indexChecksum *verify.KVChecksum
+
+	codec tikv.Codec
+}
+
+func newDeliverKVBatch(codec tikv.Codec) *deliverKVBatch {
+	return &deliverKVBatch{
+		dataChecksum:  verify.NewKVChecksumWithKeyspace(codec),
+		indexChecksum: verify.NewKVChecksumWithKeyspace(codec),
+		codec:         codec,
+	}
+}
+
+func (b *deliverKVBatch) reset() {
+	b.dataKVs.Clear()
+	b.indexKVs.Clear()
+	b.dataChecksum = verify.NewKVChecksumWithKeyspace(b.codec)
+	b.indexChecksum = verify.NewKVChecksumWithKeyspace(b.codec)
+}
+
+func (b *deliverKVBatch) size() uint64 {
+	return b.dataChecksum.SumSize() + b.indexChecksum.SumSize()
+}
+
+func (b *deliverKVBatch) add(kvs *kv.Pairs) {
+	for _, pair := range kvs.Pairs {
+		if pair.Key[tablecodec.TableSplitKeyLen+1] == 'r' {
+			b.dataKVs.Pairs = append(b.dataKVs.Pairs, pair)
+			b.dataChecksum.UpdateOne(pair)
+		} else {
+			b.indexKVs.Pairs = append(b.indexKVs.Pairs, pair)
+			b.indexChecksum.UpdateOne(pair)
+		}
+	}
+
+	// the related buf is shared, so we only need to set it into one of the kvs so it can be released
+	if kvs.BytesBuf != nil {
+		b.dataKVs.BytesBuf = kvs.BytesBuf
+		b.dataKVs.MemBuf = kvs.MemBuf
+	}
+}
+
+func firstErr(errors ...error) error {
+	for _, err := range errors {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// chunkProcessor process data chunk, it encodes and writes KV to local disk.
+type chunkProcessor struct {
+	parser      mydump.Parser
+	chunkInfo   *checkpoints.ChunkCheckpoint
+	logger      *zap.Logger
+	kvsCh       chan []deliveredRow
+	dataWriter  *backend.LocalEngineWriter
+	indexWriter *backend.LocalEngineWriter
+
+	checksum verify.KVChecksum
+	encoder  kvEncoder
+	kvCodec  tikv.Codec
+}
+
+func (p *chunkProcessor) process(ctx context.Context) error {
+	deliverCompleteCh := make(chan deliverResult)
+	go func() {
+		defer close(deliverCompleteCh)
+		err := p.deliverLoop(ctx)
```

Comment:
No need to change, but I prefer the error group pattern. When the job is finished, close the channel to notify the consumer. When any loop meets error, error group will automatically cancel the derived context and the error can be found by eg.Wait()

Reply:
simplified from lightning code, will add a comment and do it later